src/JustSaying/Messaging/Channels/Receive/MessageReceiveBuffer.cs

```diff
@@ -101,12 +102,16 @@ public async Task RunAsync(CancellationToken stoppingToken)
                         }
                     }
 
+                    if (_logger.IsEnabled(LogLevel.Trace))
+                    {
+                        _logger.LogTrace("Downloaded {MessageCount} messages from queue {QueueName}.", messages.Count, _sqsQueueReader.QueueName);
+                    }
+
                     foreach (Message message in messages)
                     {
                         IQueueMessageContext messageContext = _sqsQueueReader.ToMessageContext(message);
 
-                        // Complete all messages in the batch, rather than observing the CancellationToken to stop
-                        await writer.WriteAsync(messageContext, CancellationToken.None).ConfigureAwait(false);
+                        await writer.WriteAsync(messageContext, stoppingToken).ConfigureAwait(false);
```

Comment:
Passing CancellationToken.None here was causing a deadlock when there was no consumer attached, so this would wait forever after shutdown as the queue wasnt draining.

Reply:
Damn! Nice find